src/ray/common/ray_syncer/ray_syncer.h

```diff
@@ -0,0 +1,227 @@
+                                                 RayComponentId component_id) const = 0;
+  virtual ~ReporterInterface() {}
+};
+
+/// The interface for a receiver. Receiver is defined to be a module which would like
+/// to get the state of other nodes. For example, cluster resource manager.
+struct ReceiverInterface {
+  /// Interface to update a module. The module should read the `sync_message` fields and
+  /// deserialize it to update its internal state.
+  ///
+  /// \param message The message received from remote node.
+  virtual void Update(std::shared_ptr<const RaySyncMessage> message) = 0;
+
+  virtual ~ReceiverInterface() {}
+};
+
+// Forward declaration of internal structures
+class NodeState;
+class NodeSyncConnection;
+
+/// RaySyncer is an embedding service for component synchronization.
+/// All operations in this class needs to be finished GetIOContext()
+/// for thread-safety.
+/// RaySyncer is the control plane to make sure all connections eventually
+/// have the latest view of the cluster components registered.
+/// RaySyncer has two components:
+///    1. NodeSyncConnection: keeps track of the sending and receiving information
+///       and make sure not sending the information the remote node knows.
+///    2. NodeState: keeps track of the local status, similar to NodeSyncConnection,
+//        but it's for local node.
+class RaySyncer {
+ public:
+  /// Constructor of RaySyncer
+  ///
+  /// \param io_context The io context for this component.
+  /// \param node_id The id of current node.
+  RaySyncer(instrumented_io_context &io_context, const std::string &node_id);
+  ~RaySyncer();
+
+  /// Connect to a node.
+  /// TODO (iycheng): Introduce grpc channel pool and use node_id
+  /// for the connection.
+  ///
+  /// \param connection The connection to the remote node.
+  void Connect(std::unique_ptr<NodeSyncConnection> connection);
+
+  /// Connect to a node.
+  /// TODO (iycheng): Introduce grpc channel pool and use node_id
+  /// for the connection.
+  ///
+  /// \param connection The connection to the remote node.
+  void Connect(std::shared_ptr<grpc::Channel> channel);
+
+  void Disconnect(const std::string &node_id);
+
+  /// Register the components to the syncer module. Syncer will make sure eventually
+  /// it'll have a global view of the cluster.
+  ///
+  /// \param component_id The component to sync.
+  /// \param reporter The local component to be broadcasted.
+  /// \param receiver The snapshot of the component in the cluster.
+  /// \param upstream_only Only send the message to the upstream server for this
+  /// component. \param pull_from_reporter_interval_ms The frequence to pull a message
+  /// from reporter and push it to sending queue.
+  bool Register(RayComponentId component_id,
+                const ReporterInterface *reporter,
+                ReceiverInterface *receiver,
+                bool upstream_only = false,
+                int64_t pull_from_reporter_interval_ms = 100);
```

Comment:
make it a config value?

Reply:
not really. for different component they can have different report interval.

Reply:
That sounds good. I think as a library probably it's not a good idea to use config here to decouple the code with application logic. When we do integration, I'll include that.